components/raftstore-v2/src/operation/query/local.rs

```diff
@@ -2,44 +2,234 @@
+    errorpb,
+    raft_cmdpb::{CmdType, RaftCmdRequest, RaftCmdResponse},
 };
-use pd_client::BucketMeta;
 use raftstore::{
+    errors::RAFTSTORE_IS_BUSY,
     store::{
-        cmd_resp,
-        util::{self, LeaseState, RegionReadProgress, RemoteLease},
-        ReadDelegate, ReadExecutor, ReadExecutorProvider, ReadProgress, ReadResponse,
-        RegionSnapshot, RequestInspector, RequestPolicy,
+        cmd_resp, util::LeaseState, LocalReaderCore, ReadDelegate, ReadExecutor,
+        ReadExecutorProvider, RegionSnapshot, RequestInspector, RequestPolicy,
+        TLS_LOCAL_READ_METRICS,
     },
     Error, Result,
 };
-use slog::{debug, error, info, o, warn, Logger};
+use slog::{debug, Logger};
 use tikv_util::{
+    box_err,
     codec::number::decode_u64,
-    lru::LruCache,
-    time::{monotonic_raw_now, Instant, ThreadReadId},
+    time::{monotonic_raw_now, ThreadReadId},
 };
 use time::Timespec;
+use txn_types::WriteBatchFlags;
+
+use crate::{
+    fsm::StoreMeta,
+    router::{PeerMsg, QueryResult},
+    tablet::CachedTablet,
+    StoreRouter,
+};
+
+pub trait MsgRouter: Send {
+    fn send(&self, addr: u64, msg: PeerMsg) -> std::result::Result<(), TrySendError<PeerMsg>>;
+}
+
+impl<EK, ER> MsgRouter for StoreRouter<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    fn send(&self, addr: u64, msg: PeerMsg) -> std::result::Result<(), TrySendError<PeerMsg>> {
+        Router::send(self, addr, msg)
+    }
+}
 
-use crate::{fsm::StoreMeta, tablet::CachedTablet};
+#[derive(Clone)]
+pub struct LocalReader<E, C>
+where
+    E: KvEngine,
+    C: MsgRouter,
+{
+    local_reader: LocalReaderCore<CachedReadDelegate<E>, StoreMetaDelegate<E>>,
+    router: C,
+
+    logger: Logger,
+}
+
+impl<E, C> LocalReader<E, C>
+where
+    E: KvEngine,
+    C: MsgRouter,
+{
+    pub fn new(store_meta: Arc<Mutex<StoreMeta<E>>>, router: C, logger: Logger) -> Self {
+        Self {
+            local_reader: LocalReaderCore::new(StoreMetaDelegate::new(store_meta)),
+            router,
+            logger,
+        }
+    }
+
+    pub fn pre_propose_raft_command(
+        &mut self,
+        req: &RaftCmdRequest,
+    ) -> Result<Option<(CachedReadDelegate<E>, RequestPolicy)>> {
+        if let Some(delegate) = self.local_reader.validate_request(req)? {
+            let mut inspector = SnapRequestInspector {
+                delegate: &delegate,
+                logger: &self.logger,
+            };
+            match inspector.inspect(req) {
+                Ok(RequestPolicy::ReadLocal) => Ok(Some((delegate, RequestPolicy::ReadLocal))),
+                Ok(RequestPolicy::StaleRead) => Ok(Some((delegate, RequestPolicy::StaleRead))),
+                // It can not handle other policies.
+                Ok(_) => Ok(None),
+                Err(e) => Err(e),
+            }
+        } else {
+            Err(Error::RegionNotFound(req.get_header().get_region_id()))
+        }
+    }
+
+    fn try_get_snapshot(
```

Comment:
Do we only support snapshot here? It seems CmdType::Get is used in some test. Get is just a simple wrapper of snapshot through.

Reply:
Yes, we may remove CmdType::Get later in some time.